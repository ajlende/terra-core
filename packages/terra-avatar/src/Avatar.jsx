import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import styles from './Avatar.scss';

const cx = classNames.bind(styles);

const AvatarVariants = {
  FACILITY: 'facility',
  USER: 'user',
};

const propTypes = {
  /**
   * The text content that specifies the alternative text for the image.
   */
  alt: PropTypes.string,
  /**
   * The image to display.
   */
  image: PropTypes.string,
  /**
   * The initials to display.
   */
  initials: PropTypes.string,
  /**
   * The avatar variant. One of `AvatarVariants.FACILITY`, `AvatarVariants.USER`.
   */
  variant: PropTypes.oneOf([AvatarVariants.FACILITY, AvatarVariants.USER]),
};

const defaultProps = {
  variant: AvatarVariants.USER,
};

export const propsErrorMsg = 'Only one of the props: [image, initials] should be supplied.';

const Avatar = ({
  alt,
  image,
  initials,
  variant,
  ...customProps
  }) => {
  if (image && initials) {
    throw new Error(propsErrorMsg);
  }

  const attributes = Object.assign({}, customProps);

  const AvatarClassNames = cx([
    'avatar',
    attributes.className,
  ]);

  const AvatarChildClassNames = cx([
    'avatar-child',
    attributes.className,
  ]);

  const icon = (
    <div className={cx('avatar-icon')}>
      <svg className={cx(variant)} />
    </div>
  );

  const AvatarContent = () => {
    let avatarContent = null;
    if (image) {
      avatarContent = <img src={image} alt={alt} className={AvatarChildClassNames} />;
    } else if (initials) {
      avatarContent = <text className={AvatarChildClassNames} >{initials.toUpperCase()}</text>;
    } else {
      avatarContent = icon;
    }
    return (
      <circle aria-label="Avatar" {...attributes} className={AvatarClassNames} >
        {avatarContent}
      </circle>
    );
  };

  return <AvatarContent />;
};

Avatar.propTypes = propTypes;
Avatar.defaultProps = defaultProps;

export default Avatar;
