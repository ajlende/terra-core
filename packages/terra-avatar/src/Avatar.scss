/* Ignoring so that single quotes can be used in svg string */
/* stylelint-disable string-quotes */
@import '~terra-mixins';

:local {
  .avatar {
    background-color: var(--terra-avatar-background-color, rgba(0, 0, 0, 0.1));
    border: var(--terra-avatar-border, 0.07143rem solid rgb(28, 31, 33));
    border-radius: 50%;
    box-shadow: var(--terra-avatar-box-shadow, inset 0 0.07143rem 0.2143rem 0 rgba(0, 0, 0, 0.3));
    color: var(--terra-avatar-color, rgb(0, 0, 0));
    display: inline-block;
    font-size: var(--terra-avatar-font-size, 1.571rem);
    height: var(--terra-avatar-size, 3.429rem);
    line-height: var(--terra-avatar-line-height, 3.286rem);
    overflow: hidden;
    text-align: center;
    width: var(--terra-avatar-size, 3.429rem);
  }
  
  .avatar-icon {
    > svg {
      background-repeat: no-repeat;
      background-size: contain;
      height: 3.286rem;
      width: 3.286rem;
    }
    
    > .facility {
      background: var(--terra-avatar-icon-facility, inline-svg("<svg data-name='Layer_1' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 48 48'><path d='M33,20.2V12H15v8.2H9.8V48h12v-9h4.5v9h12V20.2H33z M17.2,42.8h-3V39h3V42.8z M17.2,33.8h-3V30h3V33.8z M25.5,33.8h-3V30h3 V33.8z M29.2,22.5h-3.8v3.8h-3v-3.8h-3.8v-3h3.8v-3.8h3v3.8h3.8V22.5z M33.8,42.8h-3V39h3V42.8z M33.8,33.8h-3V30h3V33.8z'/></svg>"));
    }
    
    > .user {
      background: var(--terra-avatar-icon-user, inline-svg("<svg data-name='Layer_1' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 48 48'><path d='M24,8.6c5.1,0,9.3,4.2,9.3,9.3s-4.2,9.3-9.3,9.3s-9.3-4.2-9.3-9.3C14.8,12.8,18.9,8.6,24,8.6z M40,48v-4.7 c0-7.6-6.1-13.8-13.7-13.8c0,0,0,0,0,0h-4.5C14.1,29.6,8,35.7,8,43.3c0,0,0,0,0,0V48H40z'/></svg>"));
    }
  }
}
