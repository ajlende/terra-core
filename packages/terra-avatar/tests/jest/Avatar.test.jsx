import React from 'react';
import Avatar from '../../src/Avatar';
import exampleProfilePhoto from '../assets/150x150.jpg';

describe('Avatar', () => {
  // Snapshot Tests
  it('should render user avatar when variant is user', () => {
    const avatar = <Avatar variant="user" />;
    const wrapper = render(avatar);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render facility avatar when variant is facility', () => {
    const avatar = <Avatar variant="facility" />;
    const wrapper = render(avatar);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render intials avatar when initials are passed in', () => {
    const avatar = <Avatar initials="JS" variant="user" />;
    const wrapper = render(avatar);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render image avatar when image is passed in', () => {
    const avatar = <Avatar image={exampleProfilePhoto} alt="placeholder" variant="user" />;
    const wrapper = render(avatar);
    expect(wrapper).toMatchSnapshot();
  });
});
