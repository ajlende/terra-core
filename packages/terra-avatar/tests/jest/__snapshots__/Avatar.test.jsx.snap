// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should render a default component 1`] = `<AvatarContent />`;

exports[`Avatar should render icon avatar when icon is passed in 1`] = `
<circle
  aria-label="Avatar"
  class="avatar"
>
  <svg
    aria-hidden="true"
    class="icon"
    data-name="Layer 1"
    focusable="false"
    height="1em"
    viewBox="0 0 48 48"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M24 0a11.3 11.3 0 1 1-11.3 11.3A11.35 11.35 0 0 1 24 0zm19.5 48v-5.7a16.77 16.77 0 0 0-16.8-16.8h-5.5A16.75 16.75 0 0 0 4.5 42.3V48z"
    />
  </svg>
</circle>
`;

exports[`Avatar should render image avatar when image is passed in 1`] = `
<circle
  aria-label="Avatar"
  class="avatar"
>
  <img
    alt="placeholder"
    class="avatar-child"
    src="test-file-stub"
  />
</circle>
`;

exports[`Avatar should render intials avatar when initials are passed in 1`] = `
<circle
  aria-label="Avatar"
  class="avatar"
>
  <text
    class="avatar-child"
  >
    JS
  </text>
</circle>
`;
